embeddings/model/lightning_module/question_answering.py

```diff
@@ -0,0 +1,161 @@
+
+    def __init__(
+            self,
+            model_name_or_path: T_path,
+            finetune_last_n_layers: int,
+            config_kwargs: Optional[Dict[str, Any]] = None,
+            task_model_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> None:
+
+        super().__init__(metrics=None, **task_model_kwargs if task_model_kwargs else {})
+        self.save_hyperparameters({"downstream_model_type": self.downstream_model_type.__name__})
+        self.downstream_model_type = self.downstream_model_type
+        self.config_kwargs = config_kwargs if config_kwargs else {}
+        self.target_names: Optional[List[str]] = None
+        self._init_model()
+
+    def _init_metrics(self) -> None:
+        pass
+
+    def _init_model(self) -> None:
+        self.config = AutoConfig.from_pretrained(
+            self.hparams.model_name_or_path,
+            **self.config_kwargs,
+        )
+        self.model: AutoModel = self.downstream_model_type.from_pretrained(
+            self.hparams.model_name_or_path, config=self.config
+        )
+        if self.hparams.finetune_last_n_layers > -1:
+            self.freeze_transformer(finetune_last_n_layers=self.hparams.finetune_last_n_layers)
+
+    def freeze_transformer(self, finetune_last_n_layers: int) -> None:
+        """Borrowed from clarinpl-embeddings library"""
+        if finetune_last_n_layers == 0:
+            for name, param in self.model.base_model.named_parameters():
+                param.requires_grad = False
+        else:
+            no_layers = self.model.config.num_hidden_layers
+            for name, param in self.model.base_model.named_parameters():
+                if name.startswith("embeddings"):
+                    layer = 0
+                elif name.startswith("encoder"):
+                    layer = int(name.split(".")[2])
+                elif name.startswith("pooler"):
+                    layer = sys.maxsize
+                else:
+                    raise ValueError("Parameter name not recognized when freezing transformer")
+                if layer >= (no_layers - finetune_last_n_layers):
+                    break
+                param.requires_grad = False
+
+    def setup(self, stage: Optional[str] = None) -> None:
+        """Borrowed from clarinpl-embeddings library"""
+        if stage in ("fit", None):
+            assert self.trainer is not None
+            if self.hparams.use_scheduler:
+                train_loader = self.trainer.datamodule.train_dataloader()
+                gpus = getattr(self.trainer, "gpus") if getattr(self.trainer, "gpus") else 0
+                tb_size = self.hparams.train_batch_size * max(1, gpus)
+                ab_size = tb_size * self.trainer.accumulate_grad_batches
+                self.total_steps: int = int(
+                    (len(train_loader.dataset) / ab_size) * float(self.trainer.max_epochs)
+                )
+            self._init_metrics()
+
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
```

Comment:
To refactor. Code duplication with `HuggingFaceLightningModule.` However, refactoring won't be so fast because it needs changes in the `HuggingFaceLightningModule` logic.